`programs/platform.h` contains Unicode character

**Describe the bug**

1. `programs/platform.h` contains Unicode (non-7bit ASCII) character.
2. When user try to use MSVC solution file `build/VS2017/lz4.sln`, this Unicode character causes the following warning.  And build process is aborted.

```
1>fuzzer.c
1> programs\platform.h(1,1): error C2220: the following warning is treated as an error
1> programs\platform.h(1,1): warning C4819: The file contains a character that cannot be represented in the current code page (932). Save the file in Unicode format to prevent data loss
```

**Expected behavior**

1. The source code doesn't contain Unicode character.
2. Complete MSVC build process without error.


**To Reproduce**
1. Open `lz4/build/VS2017/lz4.sln` with Visual C++ 2017 or greater.
2. Invoke `Menu > Build > Build Solution`.
3. During build process, compiler emits `warning C4819`.


**System (please complete the following information):**
 - OS: Windows 10 (System locale/language setting is Japan/Japanese)
 - Version: 20H2 (Build 19042.1083)
 - Compiler: MSVC 2019 (Version 19.29.30040)
 - Build System: Visual Studio (`msbuild`)

### Where's the Unicode character?

We can find it with [the following command](https://stackoverflow.com/a/9395552).

```sh
grep -P -n "[^\x00-\x7F]" programs/platform.h
```

It reports

```
83:#  if (defined(__APPLE__) && defined(__MACH__)) || defined(__SVR4) || defined(_AIX) || defined(__hpux) /* POSIX.1–2001 (SUSv3) conformant */ \
                                                                                                                    ^
```

Actual character is `U+2013` which is `0xe2, 0x80, 0x93` in UTF-8.

```sh
od -j 0x1080 -t x1 -c -A x -N 16 programs/platform.h
```

```
001080  29  20  2f  2a  20  50  4f  53  49  58  2e  31  e2  80  93  32
         )       /   *       P   O   S   I   X   .   1 342 200 223   2
                                                       ^^^^^^^^^^^
```


### What's the `warning C4819`, `error C2220` and `the current code page (932)`?

`error C2220` means "treat warning as an error".

[`warning C4819`](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4819),  is

> C4819 occurs when you compile an ANSI source file on a system using a codepage that can't represent all characters in the file.

In this case, it means

(1) Since there's no sign of UTF (BOM or [`/utf-8` switch](https://docs.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8) of `cl.exe`), MSVC try to read this source code as plain text which is encoded by the default text encoding of the system locale.
(2) In my case, it is [code page 932 (shift_jis)](https://docs.microsoft.com/en-us/windows/win32/Intl/code-page-identifiers).
(3) In this encoding system, `0xe2, 0x80, 0x93` is invalid.  And C4819 occurs.


### Possible solution

(1) Replace `U+2013` (`–`) with `U+002D` (`-`).
(2) Add `/utf-8` to MSVC project.  But it doesn't work older MSVC versions.
(3) Force to add "UTF-8 BOM" to all source code.  It may cause other compatibility issue especially for older compilers.

Perhaps, additional validation with `grep -P -n "[^\x00-\x7F]"` may help future contributors.

--- a/.travis.yml
+++ b/.travis.yml
@@ -231,6 +231,12 @@ matrix:
     - name: Compile OSS-Fuzz targets
       script:
         - ./ossfuzz/travisoss.sh
+        
+    # Unicode lint
+    # See https://github.com/lz4/lz4/issues/1018
+    - name: Run Unicode lint
+      script:
+        - ./tests/unicode_lint.sh
 
   allow_failures:
     - env: ALLOW_FAILURES=true

--- a/programs/platform.h
+++ b/programs/platform.h
@@ -80,7 +80,7 @@ extern "C" {
 ************************************************************** */
 #if !defined(_WIN32) && (defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__)) /* UNIX-like OS */ \
    || defined(__midipix__) || defined(__VMS))
-#  if (defined(__APPLE__) && defined(__MACH__)) || defined(__SVR4) || defined(_AIX) || defined(__hpux) /* POSIX.1–2001 (SUSv3) conformant */ \
+#  if (defined(__APPLE__) && defined(__MACH__)) || defined(__SVR4) || defined(_AIX) || defined(__hpux) /* POSIX.1-2001 (SUSv3) conformant */ \
      || defined(__DragonFly__) || defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__)  || defined(__MidnightBSD__) /* BSD distros */ \
      || defined(__HAIKU__)
 #    define PLATFORM_POSIX_VERSION 200112L

--- a/tests/unicode_lint.sh
+++ b/tests/unicode_lint.sh
@@ -1,0 +1,35 @@
+#!/bin/bash
+
+# `unicode_lint.sh' determines whether source files under the ./lib/ and ./programs/ directories
+# contain Unicode characters, and fails if any do.
+#
+# See https://github.com/lz4/lz4/issues/1018
+
+pass=true
+
+# Scan ./lib/ for Unicode in source (*.c, *.h) files
+result=$(
+	find ./lib/ -regex '.*\.\(c\|h\)$' -exec grep -P -n "[^\x00-\x7F]" {} \; -exec echo "FAIL: {}" \;
+)
+if [[ $result ]]; then
+	echo "$result"
+	pass=false
+fi
+
+# Scan ./programs/ for Unicode in source (*.c, *.h) files
+result=$(
+	find ./programs/ -regex '.*\.\(c\|h\)$' -exec grep -P -n "[^\x00-\x7F]" {} \; -exec echo "{}: FAIL" \;
+)
+if [[ $result ]]; then
+	echo "$result"
+	pass=false
+fi
+
+if [ "$pass" = true ]; then
+	echo "All tests successful."
+	echo "Result: PASS"
+	exit 0
+else
+	echo "Result: FAIL"
+	exit 1
+fi

--- a/tests/unicode_lint.sh
+++ b/tests/unicode_lint.sh
@@ -9,7 +9,7 @@ pass=true
 
 # Scan ./lib/ for Unicode in source (*.c, *.h) files
 result=$(
-	find ./lib/ -regex '.*\.\(c\|h\)$' -exec grep -P -n "[^\x00-\x7F]" {} \; -exec echo "FAIL: {}" \;
+	find ./lib/ -regex '.*\.\(c\|h\)$' -exec grep -P -n "[^\x00-\x7F]" {} \; -exec echo "{}: FAIL" \;
 )
 if [[ $result ]]; then
 	echo "$result"

